src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -541,6 +541,150 @@ .maxstack  2
   IL_0062:  br.s       IL_0064
   IL_0064:  ldloc.s    V_4
   IL_0066:  ret
+}");
+        }
+
+        [Fact]
+        public void DecimalEquality()
+        {
+            // demonstrate that pattern-matching against a decimal constant is
+            // at least as efficient as simply using ==
+            var source =
+@"using System;
+public class C
+{
+    public static void Main()
+    {
+        Console.Write(M1(1.0m));
+        Console.Write(M2(1.0m));
+        Console.Write(M1(2.0m));
+        Console.Write(M2(2.0m));
+    }
+
+    static int M1(decimal d)
+    {
+        if (M(d) is 1.0m) return 1;
+        return 0;
+    }
+
+    static int M2(decimal d)
+    {
+        if (M(d) == 1.0m) return 1;
+        return 0;
+    }
+
+    public static decimal M(decimal d)
+    {
+        return d;
+    }
+}";
+            var compilation = CreateCompilation(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics();
+            var expectedOutput = @"1100";
+            var compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+            compVerifier.VerifyIL("C.M1",
+@"{
+  // Code size       39 (0x27)
+  .maxstack  5
+  .locals init (bool V_0,
+                decimal V_1,
+                int V_2)
+  IL_0000:  nop
+  IL_0001:  ldarg.0
+  IL_0002:  call       ""decimal C.M(decimal)""
+  IL_0007:  stloc.1
```

Comment:
Looks like there's an unnecessary store and load in debug and release when we could just keep the value on the stack. Does emit not know to keep this as a stack local?

Reply:
If we had used the operand on the stack the operands would be in a different order. Emit doesn't know that operator== is symmetric.